common/src/stack/ansible/plugins/inventory/stacki.py

```diff
@@ -0,0 +1,130 @@
+# @copyright@
+# Copyright (c) 2006 - 2020 Teradata
+# All rights reserved. Stacki(r) v5.x stacki.com
+# https://github.com/Teradata/stacki/blob/master/LICENSE.txt
+# @copyright@
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+
+DOCUMENTATION = """
+    inventory: stacki
+    short_description: Returns hosts, groups, and vars from Stacki
+    description:
+        - Reads hosts info and attributes from Stacki.
+        - Returns host groups based on standard Stacki host selectors.
+        - Hosts have attributes attached as vars.
+"""
+
+EXAMPLES = """
+# Directly target hosts
+ansible backend-0-0 -m ping
+
+# Using standard host selectors
+ansible a_backend -m ping
+
+# In Ansible playbooks, you can use the selectors in the `hosts` attribute:
+- name: Ping my backends
+  hosts: a_backend
+  tasks:
+    - ping:
+
+# You can also use the --limit flag to `ansible-playbook`
+ansible-playbook --limit b_default playbook.yml
+
+# The following types of groups are created, with the UPPERCASE replaced
+# with the Stacki primitive:
+- a_APPLIANCE
+- e_ENVIRONMENT
+- o_OS
+- b_BOX
+- g_GROUP
+- r_RACK
+
+# All the attributes for a host are available as Ansible host vars:
+- name: Ping managed hosts
+  hosts: all
+  tasks:
+    - ping:
+      when: managed
+
+# To list all the inventory hosts, groups, and vars run:
+ansible-inventory --list
+"""
+
+from collections import defaultdict
+
+from ansible.errors import AnsibleError, AnsibleParserError
+from ansible.plugins.inventory import BaseInventoryPlugin
+import stack.api
+
+
+class InventoryModule(BaseInventoryPlugin):
+	NAME = "stacki"
+
+	def verify_file(self, path):
+		return path.endswith("@stacki")
+
+	def parse(self, inventory, loader, path, cache=True):
+		super(InventoryModule, self).parse(inventory, loader, "@stacki")
+
+		# Carve up the host info
+		appliances = defaultdict(list)
+		oses = defaultdict(list)
+		boxes = defaultdict(list)
+		racks = defaultdict(list)
+		environments = defaultdict(list)
+		groups = defaultdict(list)
+
+		for row in stack.api.Call("list host"):
+			# Add the host
+			self.inventory.add_host(row["host"])
+
+			# Record the groups it will belong to
+			appliances[row["appliance"]].append(row["host"])
+			oses[row["os"]].append(row["host"])
+			boxes[row["box"]].append(row["host"])
+			racks[row["rack"]].append(row["host"])
+
+			if row["environment"]:
+				environments[row["environment"]].append(row["host"])
+
+		# Get the host groups
+		for row in stack.api.Call("list host group"):
+			for group in row["groups"].split():
+				groups[group].append(row["host"])
+
+		# Add the 'colon' selector groups
+		selectors = (
+			('a', appliances),
+			('o', oses),
+			('b', boxes),
+			('r', racks),
+			('e', environments),
+			('g', groups)
+		)
+
+		for selector, hash_table in selectors:
+			for key in hash_table:
+				group = "%s_%s" % (selector, key)
```

Comment:
I think an f-string would be a bit cleaner here.

Reply:
I would love to have done that, but Ansible plugins need to support Python 2. Granted, I'm specifically wiring our Ansible to use our Python 3, but I thought it was best to follow convention.

Reply:
I believe this could also use the `"{}_{}".format(selector, key)` syntax instead, which is still python2 compatible (if I'm not mistaken).

Reply:
I could, if I used `{0}` and `{1}` because that's needed for Python 2.6 (the oldest Python Ansible supports) but I'm not going to, because `%` still works and is valid Python 3.